src/node_jsrtapi.cc

```diff
@@ -1004,67 +1047,157 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
 }
 
 napi_status napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
-  napi_destruct destructor, napi_weakref* handle) {
-  new ObjectWrapWrapper(jsObject, nativeObj, destructor);
+    napi_finalize finalize_cb, napi_ref* result) {
+  JsValueRef value = reinterpret_cast<JsValueRef>(jsObject);
 
-  if (handle != nullptr)
-  {
-    CHECK_NAPI(napi_create_weakref(e, jsObject, handle));
+  jsrtimpl::ExternalData* externalData = new jsrtimpl::ExternalData(nativeObj, finalize_cb);
+  CHECK_JSRT(JsSetExternalData(value, externalData));
+
+  if (result != nullptr) {
+    napi_create_reference(e, jsObject, 0, result);
   }
+
   return napi_ok;
 }
 
 napi_status napi_unwrap(napi_env e, napi_value jsObject, void** result) {
-  CHECK_ARG(result);
-  *result = ObjectWrapWrapper::Unwrap(jsObject);
-  return napi_ok;
+  return napi_get_value_external(e, jsObject, result);
 }
 
-napi_status napi_create_persistent(napi_env e, napi_value v, napi_persistent* result) {
+napi_status napi_create_external(napi_env e, void* data, napi_finalize finalize_cb,
+    napi_value* result) {
   CHECK_ARG(result);
-  JsValueRef value = reinterpret_cast<JsValueRef>(v);
-  if (value) {
-      CHECK_JSRT(JsAddRef(static_cast<JsRef>(value), nullptr));
-  }
-  *result = reinterpret_cast<napi_persistent>(value);
-  return napi_ok;
-}
 
-napi_status napi_release_persistent(napi_env e, napi_persistent p) {
-  JsValueRef thePersistent = reinterpret_cast<JsValueRef>(p);
-  CHECK_JSRT(JsRelease(static_cast<JsRef>(thePersistent), nullptr));
+  CHECK_JSRT(JsCreateExternalObject(
+    new jsrtimpl::ExternalData(data, finalize_cb),
+    jsrtimpl::ExternalData::Finalize,
+    reinterpret_cast<JsValueRef*>(result)));
+
   return napi_ok;
 }
 
-napi_status napi_get_persistent_value(napi_env e, napi_persistent p, napi_value* result) {
+napi_status napi_get_value_external(napi_env e, napi_value v, void** result) {
   CHECK_ARG(result);
-  JsValueRef value = reinterpret_cast<JsValueRef>(p);
-  *result = reinterpret_cast<napi_value>(value);
+
+  jsrtimpl::ExternalData* externalData;
+  CHECK_JSRT(JsGetExternalData(
+    reinterpret_cast<JsValueRef>(v),
+    reinterpret_cast<void**>(&externalData)));
+
+  *result = (externalData != nullptr ? externalData->Data() : nullptr);
+
   return napi_ok;
 }
 
-napi_status napi_create_weakref(napi_env e, napi_value v, napi_weakref* result)
-{
+// Set initial_refcount to 0 for a weak reference, >0 for a strong reference.
+napi_status napi_create_reference(napi_env e, napi_value v, int initial_refcount,
+    napi_ref* result) {
   CHECK_ARG(result);
+
   JsValueRef strongRef = reinterpret_cast<JsValueRef>(v);
   JsWeakRef weakRef = nullptr;
   CHECK_JSRT(JsCreateWeakReference(strongRef, &weakRef));
+
+  // Prevent the reference itself from being collected until it is explicitly deleted.
   CHECK_JSRT(JsAddRef(weakRef, nullptr));
```

Comment:
Right- my question was, whether you needed to create the weak reference there in the first place, or whether you could use a different scheme (for instance- when you create a reference, you allocate an internal "handle" object, and maintain a map of strong references to handles- in addition, you could install a before collect callback that would null out the strong reference in the handle object- that way, you only have to maintain ref count on the strong reference itself, and the handle object would get destroyed when you call napi_delete_reference)- does that make sense?